### ISSUE TITLE
[DEV] Release cycle protocol

### ISSUE THREAD

**Issue by aerugo** [cobudget/cobudget#616]
I am proposing that we adopt a more structured protocol for our releases.
This is my proposal. I have written it as if it was already adopted, to avoid rewriting to the present tense later. 
However, this is just a proposal and should be modified according to advice, especially from @gustavlrsn and @Powersource.

## Release schedule
We are currently in a very active development phase and in open beta. This means that we push new features to production after some rudimentary testing, but we accept that some bugs might be found after they go live. In a later phase of development we will become more conservative and test our releases more before going live.

Currently, we aim to push a new version of Cobudget to production on Thursday evening every other week. 
Every other Monday, three days before a new version goes live, the 'staging' branch is frozen for new updates, except for bug-fixes that are found while testing the release candidate. Sometimes fixing bugs takes longer than three days, and then we postpone pushing the new release candidate live. 

## Branches
We use three branches for our release cycle. 

- **main** is our active development branch. Everyone works from the assumption that they should keep their own WIP branches in sync with 'main'.
- **staging** is our release candidate branch. It gets automatically deployed to staging.cobudget.com.
- **production** is our production branch. It gets automatically deployed to cobudget.com.

## Merging
Almost all updates should be made from feature branches to 'main', and then merged immediately to 'staging' for testing.
While the 'staging' branch is frozen before a new release, new code is only pushed to 'main', except when the code fixes bugs or mistakes that are introduced with the new release. While this is happening, all updates to 'staging' should then be immediately merged to 'main'.

In some cases, if two people are working on large WIP features that touch the same code, they may want to also keep their respective branches in sync with each other.

After a new release is pushed to production, all new changes in 'main' are immediately merged to 'staging'.

Small bug fixes and minor fixes can be pushed straight to 'production', but they must then also be immediately merged to 'main', and then to 'staging' - or to 'staging' and then to 'main' if the staging branch is currently frozen.

In summary, this is the order of merging:

**If 'staging' is not frozen**
feature-branch -> main -> staging

**If 'staging' is frozen
feature-branch -> main

**If update is a small hotfix** 
feature-branch -> production -> staging -> main

**When a new version is released**
staging -> production

**After a new version is released**
main -> staging

**Comment by Powersource** [comment]
First, I would like the text to mention which person/role is responsible for doing all of this (guessing the product owner).

Second, his sounds like a good flow, but maybe for later when we're aiming to be even more stable. I think going from 1 to 3 branches directly is maybe a bit much right now, I think I'd like to try with just 2, I think that still could be a big improvement to what we currently have (i.e. pushing straight to prod). Would look something like:

Features are developed in feature branches, and merged to main. Main is automatically deployed to our staging server. On (every other?) tuesday afternoon(s) we are blocked from merging things bigger than hotfixes to main. If testing doesn't find things that need fixing, then we merge from main->production on wednesday morning, main is unfrozen, and the production branch is automatically deployed to the production server. If fixes are needed, then those are made on main after which we merge to production, maybe a day delayed, and main is unfrozen first then.

**Comment by aerugo** [comment]
> First, I would like the text to mention which person/role is responsible for doing all of this (guessing the product owner).

I think the product owner is responsible for:
 - Making changes to the protocol 
 - Deciding if a release candidate is stable enough to push to production
 - Notifying developers about when the stable branch is frozen
 - Deciding if we need longer or shorter release cycles

Developers are responsible for:
 - Keeping their dev branches in sync
 - Merging their updates according to the protocol (so, in the case of merging hotfixes, they would need to do 4 merges)

Does that make sense to you?

> Second, his sounds like a good flow, but maybe for later when we're aiming to be even more stable. I think going from 1 to 3 branches directly is maybe a bit much right now

I think we should consider this. Perhaps it is simpler, and enough for now.

What I worry about with the two branch approach is that sometimes the staging branch may stay frozen for longer than expected. Bugs may take longer than we thought to fix, and things may arise that delay the release. In these cases, we end up in a situation where we may have lots of unmerged feature branches that may even touch the same code. 

This might be a rare occurrence, but when it happens, I think we would be grateful to have the three branch approach. What do you think @Powersource?

**Comment by Powersource** [comment]
> Developers are responsible for:

Yeah now that I think about it at my last company it was the lead engineer who was responsible for that.

> This might be a rare occurrence

Yeah I think it could be hopefully rare enough to be worth the simplification. But hmm I'm not sure what the best solution is. I just don't envy the person who'd be doing all the merging back and forth, but maybe I overestimate the complexity. So maybe good with input from Gustav since he'd maybe be the one juggling this.

Maybe when we find a big bug we don't freeze but instead just delay merging to prod by a week?

**Comment by Powersource** [comment]
@aerugo I think I fold on this one for now, if you prefer your suggestion then let's try that for a bit.

**Comment by aerugo** [comment]
Gotcha @Powersource. Let's try it. 
I have added a section on responsibilities, and pushed this to CONTRIBUTING.md on staging: https://github.com/cobudget/cobudget/blob/staging/CONTRIBUTING.md

Later we can also add other guidelines there, like coding conventions, testing protocols, and commit message conventions.

@gustavlrsn - when you have time, could you set up the branches as described? Once that is done, we close this issue.

**Comment by Powersource** [comment]
this should be solved now but i guess we'll see for sure on thursday when we try a bigger deploy to prod (i deployed once successfully but it was only a small text file change)
